pkg/executor/aggregate/agg_spill.go

```diff
@@ -56,3 +322,112 @@ func (a *AggSpillDiskAction) Action(t *memory.Tracker) {
 func (*AggSpillDiskAction) GetPriority() int64 {
 	return memory.DefSpillPriority
 }
+
+// ParallelAggSpillDiskAction implements memory.ActionOnExceed for parallel HashAgg.
+type ParallelAggSpillDiskAction struct {
+	memory.BaseOOMAction
+	e           *HashAggExec
+	spillHelper *parallelHashAggSpillHelper
+	spillTimes  atomic.Int32
+}
+
+// Action set HashAggExec spill mode.
+func (p *ParallelAggSpillDiskAction) Action(t *memory.Tracker) {
+	if !p.actionImpl(t) {
+		return
+	}
+
+	if t.CheckExceed() {
+		p.triggerFallBackAction(t)
+	}
+}
+
+func (p *ParallelAggSpillDiskAction) checkRestriction(t *memory.Tracker) bool {
+	return hasEnoughDataToSpill(p.e.memTracker, t) && p.spillTimes.Load() < maxSpillTimes
+}
+
+// Return false if we should keep executing.
+func (p *ParallelAggSpillDiskAction) actionImpl(t *memory.Tracker) bool {
+	p.spillHelper.lock.mu.Lock()
+	defer p.spillHelper.lock.mu.Unlock()
+
+	if !p.spillHelper.isInPartialStage() {
```

Comment:
I think just use checkAllPartialWorkersFinished is enough?

Reply:
> I think just use checkAllPartialWorkersFinished is enough?

yes